products/views.py

```diff
@@ -1,3 +1,78 @@
-from django.shortcuts import render
+import json, math
+from django.db.models.aggregates import Aggregate
+from django.db.models.query_utils import Q
 
-# Create your views here.
+from django.http.response         import JsonResponse
+from django.views                 import View
+from django.db.models             import Count
+
+from products.models              import Product
+from orders.models                import OrderStatus
+from users.utils                  import login_decorator
+
+class ProductView(View):
+    @login_decorator
+    def get(self, request):
+        user                = request.user
+        product_type        = request.GET.get('type')
+        character           = request.GET.get('character')
+        search              = request.GET.get('search')
+        category            = request.GET.get('category')
+        page                = int(request.GET.get('page', 0))
+        page_size           = int(request.GET.get('pageSize', 0))
+        order_conditions    = []
+        annotate_conditions = {}
+        filter_conditions   = Q()
+        result              = {}
+
+        if product_type == 'new':
+            order_conditions.append('-created_at')
+
+        if product_type == 'hot':
+            annotate_conditions['count'] = Count("like__product_id")
+            order_conditions.append('-count')
+            order_conditions.append('-created_at')
+
+        if category:
+            filter_conditions |= Q(category__name__icontains=category)
+            order_conditions.append('-created_at') 
+
+        if character:
+            filter_conditions |= Q(character__name__icontains=character)
+            order_conditions.append('-created_at') 
+
+        if search:
+            filter_conditions |= Q(name__icontains=search)
+            order_conditions.append('-created_at')
+        
+        filtered_products = Product.objects.filter(
+            filter_conditions).annotate(
+                **annotate_conditions).order_by(
+                *order_conditions)
```

Comment:
```suggestion
        filtered_products = Product.objects.filter(q)
                                   .annotate(like_count=Count(like__product_id)
                                   .order_by(-'created_at')
```
1. like count는 조건에 상관없이 annotate를 해도 무방할 것으로 보입니다.
2. order_by 조건이 create_at 밖에 없으므로 마지막에 한 번 사용해 주면 좋을 것 같습니다.

Reply:
> like count는 조건에 상관없이 annotate를 해도 무방할 것으로 보입니다

어짜피 hot일때 말고 count를 사용해서 정렬을 하지 않으니 상관 없군요!

> order_by 조건이 create_at 밖에 없으므로 마지막에 한 번 사용해 주면 좋을 것 같습니다

type의 잘못된 option이 들어올 시에 안하도록 하고 싶었는데, 너무 깊은 생각이였던것 같습니다.!